File: ruby/day4/day4.rb
def valid?(passport)
  byr = passport["byr"].to_i
  iyr = passport["iyr"].to_i
  eyr = passport["eyr"].to_i
  hgt = passport["hgt"]
  hcl = passport["hcl"]
  ecl = passport["ecl"]
  pid = passport["pid"]

  return unless byr.between?(1920, 2002)
  return unless iyr.between?(2010, 2020)
  return unless eyr.between?(2020, 2030)

  if hgt.match?(/\d*cm/)
    return unless hgt[..-3].to_i.between?(150, 193)
  elsif hgt.match?(/\d*in/)
    return unless hgt[..-3].to_i.between?(59, 76)
  else
    return false
  end

  return unless hcl[0] == "#" && hcl.size == 7
  return unless hcl[1..].chars.all? { |c| %w(0 1 2 3 4 5 6 7 8 9 a b c d e f).include?(c) }
  return unless %w(amb blu brn gry grn hzl oth).include?(ecl)
  return unless pid.match?(/\d*/) && pid.size == 9
  true
end

p1 = 0
p2 = 0
passport = {}
valid_fields = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]

File.read('input').lines do |line|
  line = line.strip

  unless line == ""
    line.strip.split(" ").each do |x|
      key, val = x.split(":")
      passport[key] = val
    end
  else
    p line
    fields = passport.keys

    if valid_fields.all? { |f| fields.include?(f) }
      p1 += 1
      p2 += 1 if valid?(passport)
    end

    passport = {}
  end
end

p p1
p p2

File: haskell/day4/day4.hs
import Data.List.Split
import Data.List
import Data.Char
import Data.Sort
import Text.Read

type Passport = String
type PassportAttribute = (String, String)

passportAttributes :: Passport -> [PassportAttribute]
passportAttributes p = map (\[x, y] -> (x, y)) . map (splitOn ":") . foldl1 (++) . map (splitOn " ") . lines $ p

part1 :: [Passport] -> Int
part1 = length . filter validate1 . map passportAttributes

validate1 :: [PassportAttribute] -> Bool
validate1 attrs = required == attributes
  where required = sort ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]
        attributes = sort . filter (/= "cid") . map fst $ attrs

part2 :: [Passport] -> Int
part2 = length . filter validate2 . map passportAttributes

validate2 :: [PassportAttribute] -> Bool
validate2 attrs = validate1 attrs && all validateField attrs
  where validateField ("cid", _) = True
        validateField ("pid", xs) = (length xs) == 9 && (all isDigit xs)

        validateField ("ecl", "amb") = True
        validateField ("ecl", "blu") = True
        validateField ("ecl", "brn") = True
        validateField ("ecl", "gry") = True
        validateField ("ecl", "grn") = True
        validateField ("ecl", "hzl") = True
        validateField ("ecl", "oth") = True
        validateField ("ecl", _) = False

        validateField ("hcl", '#':xs) = (length xs == 6) && all ((flip elem) "0123456789abcdef") xs
        validateField ("hcl", _) = False

        validateField (key, val)
          | key == "byr" = n >= 1920 && n <= 2002
          | key == "iyr" = n >= 2010 && n <= 2020
          | key == "eyr" = n >= 2020 && n <= 2030
          | key == "hgt" && (isSuffixOf "cm" val) = hgtn >= 150 && hgtn <= 193
          | key == "hgt" && (isSuffixOf "in" val) = hgtn >= 59 && hgtn <= 76
          | otherwise = False
          where
                n = nval (readMaybe val :: Maybe Int)
                hgtn = nval (readMaybe (init . init $ val) :: Maybe Int)
                nval Nothing = 0
                nval (Just nn) = nn

main = do
    raw_input <- readFile "input"
    let passports = splitOn "\n\n" raw_input

    print $ part1 passports
    print $ part2 passports
